nifi-script-processors/S3RequestSplit.groovy

```diff
@@ -0,0 +1,206 @@
+    def numberOfParts = Math.floor(aS3Length.toBigInteger() / lfps)
+    if ((numberOfParts * lfps) < aS3Length.toBigInteger())
+    {
+      numberOfParts += 1
+    }
+    numberOfParts = numberOfParts.toInteger()
+    // ### Validate
+    if (numberOfParts >= 17575) {
+      throw new Exception("Number of parts to generate ${numberOfParts} exceeds allowable max ${fps}")
+    }
+    // ### Initialize attributes
+    def attrMap = [:]
+    attrMap.put('s3.bucket', aS3Bucket)
+    attrMap.put('filename', aFilename)
+    attrMap.put('s3.etag', aS3Etag)
+    attrMap.put('s3.isLatest', aS3IsLatest)
+    attrMap.put('s3.LastModified', aS3LastModified)
+    attrMap.put('s3.length', aS3Length)
+    attrMap.put('s3.storeClass', aS3StoreClass)
+    attrMap.put('s3.version', aS3Version)
+    attrMap.put('split.totalparts', (numberOfParts + 1).toString())
+    attrMap.put('split.originalfilename', aFilename)
+    // ### Iterate each part creating a new flow file
+    for(int i = 0; i<numberOfParts; i++)
+    {
+      // Determine a name for the split part
+      def splitPart = i + 1
+      def splitPartSuffix = getPartSuffix(i)
+      // Determine range start
+      def rangeStart = i * lfps
+      // Determine range length
+      def rangeLength = (Math.floor(splitPart) != Math.floor(numberOfParts) ? lfps : aS3Length.toBigInteger() - (i * lfps))
+      // Create a new flow file
+      log.info("Creating new flowfile part ${splitPart} (${splitPartSuffix}) of ${lfps} bytes starting from ${rangeStart} for ${aS3Bucket}${aFilename}")
+      nff = session.create(ff)
+      // Attribute updates
+      attrMap.put('split.part', Integer.toString(splitPart))
+      attrMap.put('split.suffix', splitPartSuffix)
+      attrMap.put('s3-object-range-start', rangeStart.toString() + 'B')
+      attrMap.put('s3-object-range-length', rangeLength.toString() + 'B')
+      nff = session.putAllAttributes(nff, attrMap)
+      // Transfer
+      session.transfer(nff, REL_SUCCESS)
+    }
+    // ### Add ZZZ file part intended to be empty
+    zzz = session.create(ff)
+    attrMap.put('split.part', (numberOfParts + 1).toString())   // could route on this to skip retrieval
+    attrMap.put('split.suffix', getPartSuffix(17575)) // filepart_zzz
+    attrMap.put('s3-object-range-start', '-1B')       // could route on this to skip retrieval
+    attrMap.put('s3-object-range-length', '-1B')      // could route on this to skip retrieval
+    zzz = session.putAllAttributes(zzz, attrMap)
+    // Transfer the ZZZ part
+    session.transfer(zzz, REL_SUCCESS)
+    // ### Remove the initiating flowfile that is now split up -----------------
+    session.remove(ff)
+  }
+} catch (e) {
+  // Log to console
+  log.error("Error splitting s3 requests: ${aS3Bucket}${aFilename}", e)
+  // Set attribute with error
+  ff = session.putAttribute(ff, 's3requestsplit.error.message', e.toString())
+  // Transfer
+  session.transfer(ff, REL_FAILURE)
```

Comment:
As redundant as it sounds I think there should be a safeguard on operations inside this catch block such as another try/catch statement, there are a few operations within the catch block that have potential exceptions which won't be caught and could effectively stop the processor.

Reply:
Actually thats a good point. For the ExecuteGroovyScript processor that these get laid down in, there is no guarantee that a "failure" relationship will exist.  By default the processor sets up for rollback and retry, so hitting this catch could log, and then fail to send to failure.

Reply:
That scenario is true of all the Groovy script based processors